docs/chapters/09-extract-configuration.md
# U0900 Extracting the Configuration

As we have already seen in the previous chapter, we split and refactor the Game class from beginning. Here is the step
where we split the configuration management from the main Game class.

The operation is quietly simple, we just have to create a dedicated class, and move all thing about configuration from
the Game class to the new Configuration one.

## A Configuration object

First let's create the class with its attributes:

```java
public class Configuration {

    private ResourceBundle defaultConfig;

    public String title = "fromClassToGame";

    public int width = 320;
    public int height = 200;
    public double scale = 1.0;

    public double FPS = 60;

    public String scenes = "";
    public String defaultScene = "";
}
```

We also need a goo constructor to initialize things

## U0901 reading properties

THe configration values are from a properties file.

```java
public class Configuration {
    //...
    public Configuration(){

        defaultConfig=ResourceBundle.getBundle("config");
        readValuesFromFile();
        }
    //...
}
```

And need to populate, as before, the attributes with the default values.

```java
public class Configuration {
    //...
    public void readValuesFromFile() {

        this.width = Integer.parseInt(defaultConfig.getString("game.setup.width"));
        this.height = Integer.parseInt(defaultConfig.getString("game.setup.height"));
        this.scale = Double.parseDouble(defaultConfig.getString("game.setup.scale"));
        this.FPS = Double.parseDouble(defaultConfig.getString("game.setup.fps"));
        this.title = defaultConfig.getString("game.setup.title");
        this.scenes = defaultConfig.getString("game.setup.scenes");
        this.defaultScene = defaultConfig.getString("game.setup.scene.default");
    }
    //...
}
```

## U902 Parsing arguments

And Finally, we will parse the java command line arguments to extract possible user values:

```java
class Configuration {
    public Configuration parseArgs(String[] argv) throws UnknownArgumentException {
        if (argv != null) {
            for (String arg : argv) {
                String[] values = arg.split("=");
                switch (values[0].toLowerCase()) {
                    case "width":
                        this.width = Integer.parseInt(values[1]);
                        break;
                    case "height":
                        this.height = Integer.parseInt(values[1]);
                        break;
                    case "scale":
                        this.scale = Double.parseDouble(values[1]);
                        break;
                    case "fps":
                        this.FPS = Double.parseDouble(values[1]);
                        break;
                    case "title":
                        this.title = values[1];
                        break;
                    case "scene":
                        this.defaultScene = values[1];
                        break;
                    default:
                        throw new UnknownArgumentException(String.format("The argument %s is unknown", arg));
                }
            }
        }
        return this;
    }
}
```

### U0903 Wrong argument

In case of wrong argument, an `UnknownArgumentException` is thrown.

### Updating objects using configuration

And the last but not least, Add a `configuration` attributes to the `Game` class, and the famous `getConfiguration()`,
to let other class access those values:

```java
public class Game implements KeyListener {

    private static final Logger logger = LoggerFactory.getLogger(Game.class);
    //...
    private Configuration configuration;
    //...

    public Configuration getConfiguration() {
        return this.configuration;
    }
}
```

And typically from the SceneManager, change the right line :

```java
public class SceneManager {
    //...

    public void activate() {
        activate(game.getConfiguration().defaultScene);
    }
    //...
}
```

You now have a super configuration class you may have to update to add new things.

## A possible enhancement

If you really want to get an 'infinite' version of the configuration system, you can also move all attributes to a Map,
and dynamically configure those attributes at Start, and then get there values through the Map get.

The good solution will be to implement an internal HashMap for type dedicated parser according to the attributes you
want to man,age into your configuration.

Reading the `config.property` file into
a [`ResourceBundle`](https://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html "see the official ResourceBundle Java documentation")
, then apply some Integer, Double, String or Boolean parser to each of the values to retrieve default configured one,
and store them into a Value map.

```java
public class Configuration {
    private ResourceBundle default;
    private Map<String, Object> values;

    // load default values from configuration file
    Configuration(String configurationPath) {
        default =ResourceBundle.getBundle(configurationPath);
        values = new HashMap<>();
        readDefaultValues();
    }

    // parse the Java CLI arguments
    public void parseArgs(String[] argv) {
        //...
    }

    // extract default values from Configuration file
    public void readDefaultValues() {...}

    // configuration value accessor.
    public String getString(String name) {
        return values.get(name);
    }

    public Boolean getBoolean(String name) {
        return Boolean.parseBoolean(values.get(name));
    }

    public Integer getInteger(String name) {
        return Integer.parseInt(values.get(name));
    }

    public Double getDouble(String name) {
        return Double.parseDouble(values.get(name));
    }

    public String[] getArrayString(String name) {
        return values.get("name").split(',');
    }
}
```

## Evolution

The Configuration class will integrate progressively more and more parameters to be set and loaded from the properties
file.

The best way to add new attributes to the configuration will be to add some helpers on definition and on parsing vaue,
but also on help side. Requesting from the CLI, you would get a well formated and documented help bout possible
parameters and their values.

You would get some `ArgParser` interface and a `CLIManager` to support those parser.

### the IArgParser interface

```java
public interface IArgParser<T> {
    public boolean validate(String strValue);

    public T getValue();

    public String getShortKey();

    public String getLongKey();

    public String getName();

    public String getDescription();

    public String getErrorMessage(Object[] args);

    public T getDefaultValue();
}
```

In this interface, the main intersting thing are name and description, useful to generate the help on the cli. Name will
be the `name` of the parameter, and `description` will help understanding its usage.

An abstract class will provide a default implmentation for all parsers.

```java
public abstract class ArgParser<T> implements IArgParser<T> {

    public String name;
    public String shortKey;
    public String longKey;
    public Class<?> type;
    public T value;
    public T defaultValue;
    public String description;
    public String errorMessage;

    protected ArgParser() {

    }

    protected ArgParser(String name, String shortKey, String longKey, String description) {
        this.name = name;
        this.shortKey = shortKey;
        this.longKey = longKey;
        this.defaultValue = defaultValue;
        this.description = description;
    }

    public abstract T parse(String strValue);

    public abstract boolean validate(String strValue);

    /**
     * @return the longKey
     */
    public String getLongKey() {
        return longKey;
    }

    @Override
    public T getDefaultValue() {
        return defaultValue;
    }

    @Override
    public String getDescription() {
        return String.format("[%s/%s] : %s ( default:%s )", shortKey, longKey, description, defaultValue);
    }

    @Override
    public String getErrorMessage(Object[] args) {
        return String.format(errorMessage, args);
    }

    @Override
    public String getShortKey() {
        return shortKey;
    }

    @Override
    public String getName() {
        return name;
    }

    public T getValue() {
        return value;
    }
}
```

And the CLIManager will be used to parse command line interface parameters, but also to provide the help text.

```java
public class CliManager {
    @SuppressWarnings("unused")
    private Game game;
    private Map<String, IArgParser<?>> argParsers = new HashMap<>();
    private Map<String, Object> values = new HashMap<>();

    public CliManager(Game g) {
        this.game = g;
    }

    public void add(IArgParser<?> ap) {
        argParsers.put(ap.getName(), ap);
        log.debug("add cli parser for " + ap.getDescription());
    }

    public void parse(String[] args) {
        for (String arg : args) {
            if (arg.equals("h") || arg.equals("help")) {
                System.out.println("\n\nCommand Usage:\n--------------");
                for (IArgParser<?> ap : argParsers.values()) {
                    System.out.println("- " + ap.getDescription());
                }
                System.exit(0);
            } else {
                parseArguments(arg);
            }
        }
    }

    private void parseArguments(String arg) {
        String[] itemValue = arg.split("=");
        for (IArgParser<?> ap : argParsers.values()) {
            if (ap.getShortKey().equals(itemValue[0]) || ap.getLongKey().equals(itemValue[0])) {
                if (ap.validate(itemValue[1])) {
                    values.put(ap.getName(), ap.getValue());
                } else {
                    log.error(ap.getErrorMessage(null));
                }
            }
        }
    }

    public Object getValue(String key) throws ArgumentUnknownException {
        if (values.containsKey(key)) {
            return values.get(key);
        } else {
            return argParsers.get(key).getDefaultValue();
        }
    }

    public boolean isExists(String key) {
        return values.containsKey(key);
    }
}
```

### Implementing a parser

An Integer parameter will have to parse int values.

```java
public class IntArgParser extends ArgParser<Integer> {

    public IntArgParser() {
        super();
    }

    public IntArgParser(
            String name,
            String shortKey,
            String longKey,
            int defaultValue,
            int min,
            int max,
            String description,
            String errorMessage) {
        super(name, shortKey, longKey, defaultValue, min, max, description, errorMessage);
    }

    @Override
    public boolean validate(String strValue) {
        value = defaultValue;
        try {
            value = parse(strValue);
            if ((min != null && value < min) || (max != null && value > max)) {
                errorMessage += String.format(
                        "value for %s must be between %s and %s. Value has been limited to min/max", name, min, max,
                        defaultValue);
                value = (value < min ? min : (value > max ? max : value));
            }
        } catch (Exception e) {
            value = defaultValue;
            errorMessage += String.format("value %s for argument %s is not possible.reset to default Value %s",
                    strValue, name, defaultValue);
            return false;
        }
        return true;
    }

    @Override
    public Integer parse(String strValue) {
        6
        int value = Integer.parseInt(strValue);
        return value;
    }
}
```

The same way to implement Boolean, Double and Float will be used.

- `BooleanArgParser` will parse a boolean value parameter,
- `DoubleArgParser` will parse a double value parameter,
- `FloatArgParser` will parse a float value parameter.

And specific one will be

- `IntArrayArgParser` will parse a list of int values parameter.

And a final pass would provide a way to load/save those parameters to a configuration file.

But this is another story !


src/docs/chapters/09-extract-configuration.md
---
title: From a Class to Game
chapter: 09 - Extracting the Configuration
author: Frédéric Delorme
description: A dedicated service to Scenes management.
created: 2021-08-01
tags: gamedev, scene
---

## Extracting the Configuration

As we have already seen in the previous chapter, we split and refactor the Game class from beginning. Here is the step where
we split the configuration management from the main Game class.

The operation is quietly simple, we just have to create a dedicated class, and move all thing about configuration from
the Game class to the new Configuration one.

First let's create the class with its attributes:

```java
public class Configuration {

    private ResourceBundle defaultConfig;

    public String title = "fromClassToGame";

    public int width = 320;
    public int height = 200;
    public double scale = 1.0;

    public double FPS = 60;

    public String scenes = "";
    public String defaultScene = "";
```

We also need a goo constructor to initialize things

```java
    public Configuration(){

        defaultConfig=ResourceBundle.getBundle("config");
        readValuesFromFile();
        }
```

And need to populate, as before, the attributes with the default values.

```java
    public void readValuesFromFile(){

        this.width=Integer.parseInt(defaultConfig.getString("game.setup.width"));
        this.height=Integer.parseInt(defaultConfig.getString("game.setup.height"));
        this.scale=Double.parseDouble(defaultConfig.getString("game.setup.scale"));
        this.FPS=Double.parseDouble(defaultConfig.getString("game.setup.fps"));
        this.title=defaultConfig.getString("game.setup.title");
        this.scenes=defaultConfig.getString("game.setup.scenes");
        this.defaultScene=defaultConfig.getString("game.setup.scene.default");
        }
```

And Finally, we will parse the java command line arguments to extract possible user values:

```java
class Configuration {
    public Configuration parseArgs(String[] argv) throws UnknownArgumentException {
        if (argv != null) {
            for (String arg : argv) {
                String[] values = arg.split("=");
                switch (values[0].toLowerCase()) {
                    case "width":
                        this.width = Integer.parseInt(values[1]);
                        break;
                    case "height":
                        this.height = Integer.parseInt(values[1]);
                        break;
                    case "scale":
                        this.scale = Double.parseDouble(values[1]);
                        break;
                    case "fps":
                        this.FPS = Double.parseDouble(values[1]);
                        break;
                    case "title":
                        this.title = values[1];
                        break;
                    case "scene":
                        this.defaultScene = values[1];
                        break;
                    default:
                        throw new UnknownArgumentException(String.format("The argument %s is unknown", arg));
                }
            }
        }
        return this;
    }
}
```

And the last but not least, Add a `configuration` attributes to the `Game` class, and the famous `getConfiguration()`,
to let other class access those values:

```java
public class Game implements KeyListener {

    private static final Logger logger = LoggerFactory.getLogger(Game.class);
    ...
    private Configuration configuration;
    ...

    public Configuration getConfiguration() {
        return this.configuration;
    }
}
```

And typically from the SceneManager, change the right line :

```java
public class SceneManager {
    ...

    public void activate() {
        activate(game.getConfiguration().defaultScene);
    }
    ...
}
```

You now have a super configuration class you may have to update to add new things.

### A possible enhancement

If you really want to get an 'infinite' version of the configuration system, you can also move all attributes to a Map,
and dynamically configure those attributes at Start, and then get there values through the Map get.

The good solution will be to implement an internal HashMap for type dedicated parser according to the attributes you
want to man,age into your configuration.

Reading the `config.property` file into
a [`ResourceBundle`](https://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html "see the official ResourceBundle Java documentation")
, then apply some Integer, Double, String or Boolean parser to each of the values to retrieve default configured one,
and store them into a Value map.

```java
public class Configuration {
    private ResourceBundle default;
    private Map<String, Object> values;

    // load default values from configuration file
    Configuration(String configurationPath) {
        default = ResourceBundle.getBundle(configurationPath);
        values = new HashMap<>();
        readDefaultValues();
    }

    // parse the Java CLI arguments
    public void parseArgs(String[] argv) {
        ...
    }

    // extract default values from Configuration file
    public void readDefaultValues() {...}

    // configuration value accessor.
    public String getString(String name) {
        return values.get(name);
    }

    public Boolean getBoolean(String name) {
        return Boolean.parseBoolean(values.get(name));
    }

    public Integer getInteger(String name) {
        return Integer.parseInt(values.get(name));
    }

    public Double getDouble(String name) {
        return Double.parseDouble(values.get(name));
    }

    public String[] getArrayString(String name) {
        return values.get("name").split(',');
    }
}
```

### Evolution

The Configuration class will integrate progressively more and
more parameters to be set and loaded from the properties file.

The best way to add new attributes to the configuration will be to add
some helpers on definition and on parsing vaue, but also on help side.
Requesting from the CLI, you would get a well formated and documented
help bout possible parameters and their values.

You would get some `ArgParser` interface and a `CLIManager` to support those parser.

#### the IArgParser interface

```java
public interface IArgParser<T> {
    public boolean validate(String strValue);
    public T getValue();
    public String getShortKey();
    public String getLongKey();
    public String getName();
    public String getDescription();
    public String getErrorMessage(Object[] args);
    public T getDefaultValue();
}
```

In this interface, the main intersting thing are name and description,
useful to generate the help on the cli. Name will be the `name` of the parameter, and `description` will help understanding its usage.

An abstract class will provide a default implmentation for all parsers.

```java
public abstract class ArgParser<T> implements IArgParser<T> {

    public String name;
    public String shortKey;
    public String longKey;
    public Class<?> type;
    public T value;
    public T defaultValue;
    public String description;
    public String errorMessage;

    protected ArgParser() {

    }

    protected ArgParser(String name, String shortKey, String longKey, String description) {
        this.name = name;
        this.shortKey = shortKey;
        this.longKey = longKey;
        this.defaultValue = defaultValue;
        this.description = description;
    }

    public abstract T parse(String strValue);

    public abstract boolean validate(String strValue);

    /**
     * @return the longKey
     */
    public String getLongKey() {
        return longKey;
    }

    @Override
    public T getDefaultValue() {
        return defaultValue;
    }

    @Override
    public String getDescription() {
        return String.format("[%s/%s] : %s ( default:%s )",shortKey,longKey,description,defaultValue);
    }

    @Override
    public String getErrorMessage(Object[] args) {
        return String.format(errorMessage, args);
    }

    @Override
    public String getShortKey() {
        return shortKey;
    }

    @Override
    public String getName() {
        return name;
    }

    public T getValue(){
        return value;
    }
}
```

And the CLIManager will be used to parse command line interface parameters, but also to provide the help text.

```java
public class CliManager {
  @SuppressWarnings("unused")
  private Game game;
  private Map<String, IArgParser<?>> argParsers = new HashMap<>();
  private Map<String, Object> values = new HashMap<>();
  public CliManager(Game g) {
    this.game = g;
  }
  public void add(IArgParser<?> ap) {
    argParsers.put(ap.getName(), ap);
    log.debug("add cli parser for " + ap.getDescription());
  }
  public void parse(String[] args) {
    for (String arg : args) {
      if (arg.equals("h") || arg.equals("help")) {
        System.out.println("\n\nCommand Usage:\n--------------");
        for (IArgParser<?> ap : argParsers.values()) {
          System.out.println("- " + ap.getDescription());
        }
        System.exit(0);
      } else {
        parseArguments(arg);
      }
    }
  }
  private void parseArguments(String arg) {
    String[] itemValue = arg.split("=");
    for (IArgParser<?> ap : argParsers.values()) {
      if (ap.getShortKey().equals(itemValue[0]) || ap.getLongKey().equals(itemValue[0])) {
        if (ap.validate(itemValue[1])) {
          values.put(ap.getName(), ap.getValue());
        } else {
          log.error(ap.getErrorMessage(null));
        }
      }
    }
  }
  public Object getValue(String key) throws ArgumentUnknownException {
    if (values.containsKey(key)) {
      return values.get(key);
    } else {
      return argParsers.get(key).getDefaultValue();
    }
  }
  public boolean isExists(String key) {
    return values.containsKey(key);
  }
}
```

#### Implementing a parser

An Integer parameter will have to parse int values.

```java
public class IntArgParser extends ArgParser<Integer>{

    public IntArgParser() {
        super();
    }
    public IntArgParser(
        String name, 
        String shortKey, 
        String longKey, 
        int defaultValue, 
        int min, 
        int max,
        String description, 
        String errorMessage) {
            super(name, shortKey, longKey, defaultValue, min, max, description, errorMessage);
    }
    @Override
    public boolean validate(String strValue) {
        value = defaultValue;
        try {
            value = parse(strValue);
            if ((min != null && value < min) || (max != null && value > max)) {
                errorMessage += String.format(
                        "value for %s must be between %s and %s. Value has been limited to min/max", name, min, max,
                        defaultValue);
                value = (value < min ? min : (value > max ? max : value));
            }
        } catch (Exception e) {
            value = defaultValue;
            errorMessage += String.format("value %s for argument %s is not possible.reset to default Value %s",
                    strValue, name, defaultValue);
            return false;
        }
        return true;
    }
    @Override
    public Integer parse(String strValue) {6
        int value  = Integer.parseInt(strValue);
        return value;
    }
}
```

The same way to implement Boolean, Double and Float will be used.

- `BooleanArgParser` will parse a boolean value parameter,
- `DoubleArgParser` will parse a double value parameter,
- `FloatArgParser` will parse a float value parameter.

And specific one will be

- `IntArrayArgParser` will parse a list of int values parameter.9

And a final pass would provide a way to load/save those parameters to a configuration file.

But this is another story !


src/main/java/fr/snapgames/fromclasstogame/core/config/Configuration.java
package fr.snapgames.fromclasstogame.core.config;

import java.util.ResourceBundle;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import fr.snapgames.fromclasstogame.core.config.cli.BooleanArgParser;
import fr.snapgames.fromclasstogame.core.config.cli.CliManager;
import fr.snapgames.fromclasstogame.core.config.cli.DoubleArgParser;
import fr.snapgames.fromclasstogame.core.config.cli.FloatArgParser;
import fr.snapgames.fromclasstogame.core.config.cli.IntegerArgParser;
import fr.snapgames.fromclasstogame.core.config.cli.StringArgParser;
import fr.snapgames.fromclasstogame.core.config.cli.Vector2dArgParser;
import fr.snapgames.fromclasstogame.core.physic.Vector2d;

/**
 * THe Game Configuration implementation of {@link AbstractConfiguration} will defined specific needs for our Game.
 *
 * @author Frédéric Delorme
 * @see AbstractConfiguration
 * @since 0.0.1
 */
public class Configuration extends AbstractConfiguration {
    /**
     * path to the default configuration file
     */
    private static final String CFG_KEY_CONFIG = "config";

    /**
     * value of the default debug level (0 to 5 fixing the detailed granularity,
     * default must be set to 0 for normal mode)
     */
    private static final String CFG_KEY_DEBUG = "debug";

    /**
     * default title of the window.
     */
    private static final String CFG_KEY_TITLE = "title";

    /**
     * default scene code to be activated
     */
    private static final String CFG_KEY_SCENE = "scene";

    /**
     * list of scenes
     */
    private static final String CFG_KEY_SCENES = "scenes";

    /**
     * destination display to show the game window
     */
    private static final String CFG_KEY_DISPLAY = "display";


    /**
     * Frame Per Second to set display framerate.
     */
    private static final String CFG_KEY_FPS = "FPS";

    /**
     * pixel scale of the game rendering
     */
    private static final String CFG_KEY_SCALE = "scale";

    /**
     * default height of the window.
     */
    private static final String CFG_KEY_HEIGHT = "height";

    /**
     * default width of the window.
     */
    private static final String CFG_KEY_WIDTH = "width";

    /**
     * Audio mute parameter.
     */
    private static final String CFG_KEY_AUDIO_MUTE = "mute";

    /**
     * Audio Sound Volume
     */
    private static final String CFG_KEY_AUDIO_SOUND_VOLUME = "sound";

    /**
     * Audio Sound Volume
     */
    private static final String CFG_KEY_AUDIO_MUSIC_VOLUME = "music";


    /**
     * vector of the default gravty for the physic Engine
     */
    private static final String CFG_KEY_WORLD_GRAVITY = "gravity";
    /**
     * Deign the world width at start.
     */
    private static final String CFG_KEY_WORLD_WIDTH = "worldWidth";
    /**
     * Deign the world width at start.
     */
    private static final String CFG_KEY_WORLD_HEIGHT = "worldHeight";


    private static final Logger logger = LoggerFactory.getLogger(Configuration.class);

    public String title = "fromClassToGame";

    /**
     * Display parameters
     */
    public int width = 320;
    public int height = 200;
    public double scale = 1.0;
    public double FPS = 60;
    public int defaultScreen = 0;

    /**
     * world parameters
     */
    public int worldWidth;
    public int worldHeight;
    public Vector2d gravity = new Vector2d(0.0, -0.981);

    public String scenes = "";
    public String defaultScene = "";

    /**
     * Tilemap and Level root path
     */
    public String levelPath;


    /**
     * Audio Parameters
     */
    public boolean mute;
    public float soundVolume;
    public float musicVolume;

    /**
     * Debug level
     */
    public int debugLevel;


    public Configuration(String configurationPath) {
        super(configurationPath);
    }

    @Override
    public void initializeArgParser(String configurationPath) {

        /*---- Debug level mode activation ----*/
        cm.add(new IntegerArgParser(CFG_KEY_DEBUG, "dbg", CFG_KEY_DEBUG,
                "set the value for the debug mode (0 to 5)", "game.setup.debugLevel", 0));

        /*---- Game window title ----*/
        cm.add(new StringArgParser(CFG_KEY_TITLE, "t", CFG_KEY_TITLE, "Define the game window title",
                "game.setup.title", CFG_KEY_TITLE));

        /*---- Display screen setup ----*/
        cm.add(new IntegerArgParser(CFG_KEY_WIDTH, "w", CFG_KEY_WIDTH, "default width of the game screen",
                "game.setup.width", 320));
        cm.add(new IntegerArgParser(CFG_KEY_HEIGHT, "h", CFG_KEY_HEIGHT, "default height of the game screen",
                "game.setup.height", 200));
        cm.add(new DoubleArgParser(CFG_KEY_SCALE, "sc", CFG_KEY_SCALE, "default game screen scaling",
                "game.setup.scale", 2.0));
        cm.add(new IntegerArgParser(CFG_KEY_FPS, "f", "fps", "set the frame per second (25-60)",
                "game.setup.fps", 60));
        cm.add(new IntegerArgParser(CFG_KEY_DISPLAY, "di", CFG_KEY_DISPLAY,
                "set the default display to play the game", "game.setup.screen", 0));

        /*---- Scene setup ----*/
        cm.add(new StringArgParser(CFG_KEY_SCENES, "ss", CFG_KEY_SCENES,
                "Define the scene names and classes to initialize the game", "game.setup.scenes.list",
                ""));
        cm.add(new StringArgParser(CFG_KEY_SCENE, "sd", CFG_KEY_SCENE, "Define the default scene to start with",
                "game.setup.scenes.default", ""));
        cm.add(new StringArgParser(CFG_KEY_CONFIG, "c", CFG_KEY_CONFIG,
                "set the path and file to be loaded for configuration", "game.setup.config.filename",
                configurationPath));

        /*---- Parameters for Audio service ----*/
        cm.add(new BooleanArgParser(CFG_KEY_AUDIO_MUTE, "mute", CFG_KEY_AUDIO_MUTE,
                "Define the audio mute flag", "game.audio.mote.flag", false));
        cm.add(new FloatArgParser(CFG_KEY_AUDIO_SOUND_VOLUME, "snd", CFG_KEY_AUDIO_SOUND_VOLUME,
                "Define the audio volume level (0.0 to 1.0)", "game.audio.volume.sound", 1.0f));
        cm.add(new FloatArgParser(CFG_KEY_AUDIO_MUSIC_VOLUME, "mus", CFG_KEY_AUDIO_SOUND_VOLUME,
                "Define the music volume level (0.0 to 1.0)",
                "game.audio.volume.music", 0.8f));

        /*---- Parameters for World setup ----*/
        cm.add(new Vector2dArgParser(CFG_KEY_WORLD_GRAVITY, "g", CFG_KEY_WORLD_GRAVITY, "define the default game gravity",
                "game.setup.world.gravity", new Vector2d(0, -0.981)));
        cm.add(new IntegerArgParser(CFG_KEY_WORLD_WIDTH, "ww", CFG_KEY_WORLD_WIDTH,
                "Define the world width (in pixel, max 2000)",
                "game.setup.world.width", 800));
        cm.add(new IntegerArgParser(CFG_KEY_WORLD_HEIGHT, "ww", CFG_KEY_WORLD_HEIGHT,
                "Define the world heiht (in pixel, max 2000)",
                "game.setup.world.height", 800));
    }


    @Override
    public void getValuesFromCM() {
        this.debugLevel = (Integer) cm.getValue(CFG_KEY_DEBUG);
        this.title = (String) cm.getValue(CFG_KEY_TITLE);

        this.width = (Integer) cm.getValue(CFG_KEY_WIDTH);
        this.height = (Integer) cm.getValue(CFG_KEY_HEIGHT);
        this.scale = (Double) cm.getValue(CFG_KEY_SCALE);

        this.defaultScreen = (Integer) cm.getValue(CFG_KEY_DISPLAY);
        this.FPS = (Integer) cm.getValue(CFG_KEY_FPS);

        this.defaultScene = (String) cm.getValue(CFG_KEY_SCENE);
        this.scenes = (String) cm.getValue(CFG_KEY_SCENES);

        this.configPath = (String) cm.getValue(CFG_KEY_CONFIG);

        this.mute = (Boolean) cm.getValue(CFG_KEY_AUDIO_MUTE);
        this.soundVolume = (Float) cm.getValue(CFG_KEY_AUDIO_SOUND_VOLUME);
        this.musicVolume = (Float) cm.getValue(CFG_KEY_AUDIO_MUSIC_VOLUME);

        this.worldWidth = (Integer) cm.getValue(CFG_KEY_WORLD_WIDTH);
        this.worldHeight = (Integer) cm.getValue(CFG_KEY_WORLD_HEIGHT);
        this.gravity = (Vector2d) cm.getValue(CFG_KEY_WORLD_GRAVITY);

    }
}
